// metaapi.cloud-sdk/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    CopyFactory: function() {
        return _metaapicloudcopyfactorysdk.default;
    },
    EquityBalanceListener: function() {
        return _riskManagement.EquityBalanceListener;
    },
    EquityChartListener: function() {
        return _riskManagement.EquityChartListener;
    },
    HistoryStorage: function() {
        return _historyStorage.default;
    },
    MemoryHistoryStorage: function() {
        return _memoryHistoryStorage.default;
    },
    MetaStats: function() {
        return _metaapicloudmetastatssdk.default;
    },
    PeriodStatisticsListener: function() {
        return _riskManagement.PeriodStatisticsListener;
    },
    RiskManagement: function() {
        return _riskManagement.default;
    },
    StopoutListener: function() {
        return _metaapicloudcopyfactorysdk.StopoutListener;
    },
    SynchronizationListener: function() {
        return _synchronizationListener.default;
    },
    TrackerEventListener: function() {
        return _riskManagement.TrackerEventListener;
    },
    TransactionListener: function() {
        return _metaapicloudcopyfactorysdk.TransactionListener;
    },
    UserLogListener: function() {
        return _metaapicloudcopyfactorysdk.UserLogListener;
    },
    default: function() {
        return _default;
    }
});
const _metaApi = /*#__PURE__*/ _interop_require_default(require("./metaApi/metaApi"));
const _historyStorage = /*#__PURE__*/ _interop_require_default(require("./metaApi/historyStorage"));
const _memoryHistoryStorage = /*#__PURE__*/ _interop_require_default(require("./metaApi/memoryHistoryStorage"));
const _synchronizationListener = /*#__PURE__*/ _interop_require_default(require("./clients/metaApi/synchronizationListener"));
const _metaapicloudcopyfactorysdk = /*#__PURE__*/ _interop_require_wildcard(require("metaapi.cloud-copyfactory-sdk"));
const _metaapicloudmetastatssdk = /*#__PURE__*/ _interop_require_default(require("metaapi.cloud-metastats-sdk"));
const _riskManagement = /*#__PURE__*/ _interop_require_wildcard(require("./riskManagement"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
console.log("MetaApi SDK loaded");
const _default = _metaApi.default;
